// ng2-rut/__ivy_ngcc__/dist/ng2-rut.module.js
"use strict";
var core_1 = require('@angular/core');
var forms_1 = require('@angular/forms');
var rut_pipe_1 = require('./rut.pipe');
var rut_validator_1 = require('./rut.validator');
var rut_directive_1 = require('./rut.directive');
var rut_value_accessor_1 = require('./rut-value-accessor');
var rut_pipe_2 = require('./rut.pipe');
exports.RutPipe = rut_pipe_2.RutPipe;
var rut_validator_2 = require('./rut.validator');
exports.RutValidator = rut_validator_2.RutValidator;
var rut_directive_2 = require('./rut.directive');
exports.RutDirective = rut_directive_2.RutDirective;
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./rut.pipe');
var ɵngcc2 = require('./rut.directive');
var ɵngcc3 = require('./rut.validator');
var ɵngcc4 = require('./rut-value-accessor');
var ɵngcc5 = require('@angular/forms');
var Ng2Rut = (function () {
    function Ng2Rut() {
    }
    /** @nocollapse */
    Ng2Rut.ctorParameters = function () { return []; };
Ng2Rut.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2Rut });
Ng2Rut.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng2Rut_Factory(t) { return new (t || Ng2Rut)(); }, providers: [
        rut_validator_1.RutValidator,
    ], imports: [[
            forms_1.FormsModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2Rut, { declarations: [ɵngcc1.RutPipe, ɵngcc2.RutDirective, ɵngcc3.RutValidator, ɵngcc4.RutValueAccessor], imports: [ɵngcc5.FormsModule], exports: [ɵngcc1.RutPipe, ɵngcc2.RutDirective, ɵngcc3.RutValidator, ɵngcc4.RutValueAccessor] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng2Rut, [{
        type: core_1.NgModule,
        args: [{
                imports: [
                    forms_1.FormsModule,
                ],
                declarations: [
                    rut_pipe_1.RutPipe,
                    rut_directive_1.RutDirective,
                    rut_validator_1.RutValidator,
                    rut_value_accessor_1.RutValueAccessor,
                ],
                providers: [
                    rut_validator_1.RutValidator,
                ],
                exports: [
                    rut_pipe_1.RutPipe,
                    rut_directive_1.RutDirective,
                    rut_validator_1.RutValidator,
                    rut_value_accessor_1.RutValueAccessor,
                ]
            }]
    }], function () { return []; }, null); })();
    return Ng2Rut;
}());
exports.Ng2Rut = Ng2Rut;

exports.RutValueAccessor = ɵngcc4.RutValueAccessor;
