// ng2-rut/__ivy_ngcc__/dist/rut.validator.js
"use strict";
var core_1 = require('@angular/core');
var forms_1 = require('@angular/forms');
var rut_helpers_1 = require('rut-helpers');
function validateRutFactory(rutValidate) {
    return function (c) {
        return rutValidate(c.value) ? null : { invalidRut: true };
    };
}
exports.validateRutFactory = validateRutFactory;
var ɵngcc0 = require('@angular/core');
var RutValidator = (function () {
    function RutValidator() {
        this.validator = validateRutFactory(rut_helpers_1.rutValidate);
    }
    RutValidator.prototype.validate = function (c) {
        return this.validator(c);
    };
    /** @nocollapse */
    RutValidator.ctorParameters = function () { return []; };
RutValidator.ɵfac = function RutValidator_Factory(t) { return new (t || RutValidator)(); };
RutValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: RutValidator, selectors: [["", "validateRut", "", "ngModel", ""], ["", "validateRut", "", "formControl", ""]], features: [ɵngcc0.ɵɵProvidersFeature([
            { provide: forms_1.NG_VALIDATORS, useExisting: core_1.forwardRef(function () { return RutValidator; }), multi: true },
        ])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RutValidator, [{
        type: core_1.Directive,
        args: [{
                selector: '[validateRut][ngModel],[validateRut][formControl]',
                providers: [
                    { provide: forms_1.NG_VALIDATORS, useExisting: core_1.forwardRef(function () { return RutValidator; }), multi: true },
                ]
            }]
    }], function () { return []; }, null); })();
    return RutValidator;
}());
exports.RutValidator = RutValidator;
