// ngx-printer/esm2015/lib/ngx-printer.service.js
import { Injectable, TemplateRef, ComponentFactoryResolver, Injector, Optional } from '@angular/core';
import { NgxPrinterComponent } from './ngx-printer.component';
import { BehaviorSubject } from 'rxjs';
import { PrintServiceConfig } from './print-service-config';
import * as i0 from "@angular/core";
import * as i1 from "./print-service-config";
/**
 * Main print service
 */
export class NgxPrinterService {
    constructor(config, resolver, injector) {
        this.resolver = resolver;
        this.injector = injector;
        this.printWindowOpen = new BehaviorSubject(false);
        /**
         * @internal
         */
        this._printItems = new BehaviorSubject([]);
        this.$printItems = this._printItems.asObservable();
        /**
         * Wait time to render before open print dialog in ms
         * Default is 200
         */
        this.timeToWaitRender = 200;
        /**
         * Class used in component when printing to current window
         */
        this.renderClass = 'default';
        /**
         * Open new window to print or not
         * Default is true
         */
        this.printOpenWindow = true;
        /**
         * Name of root component
         * Default is app-root
         */
        this.appRootName = 'app-root';
        /**
         * Do not fire print event - just show preview
         * Default is false
         */
        this.printPreviewOnly = false;
        this.appRootDislaySetting = '';
        this.$printWindowOpen = this.printWindowOpen.asObservable();
        this.eventadded = [];
        this.setRootConfigOptions(config);
    }
    /**
     * Set config from forRoot
     * @param config
     */
    setRootConfigOptions(config) {
        if (config) {
            if (config.printOpenWindow) {
                this.printOpenWindow = config.printOpenWindow;
            }
            if (config.timeToWaitRender) {
                this.timeToWaitRender = config.timeToWaitRender;
            }
            if (config.renderClass) {
                this.renderClass = config.renderClass;
            }
            if (config.appRootName) {
                this.appRootName = config.appRootName;
            }
            if (config.markerPosition) {
                this.markerPosition = config.markerPosition;
            }
            if (config.printPreviewOnly) {
                this.printPreviewOnly = config.printPreviewOnly;
            }
        }
    }
    /***
     * Print a div identified by its id
     * @example
     * this.printerService.printDiv('printDiv');
     */
    printDiv(divID) {
        const divToPrint = document.getElementById(divID);
        if (divToPrint) {
            this.print(divToPrint, this.printOpenWindow);
        }
        else {
            console.log(`div with id ${divID} not found..`);
        }
    }
    /***
     * Print an Element identified by its className using getElementsByClassName
     * Prints the first one found
     */
    printByClassName(className) {
        const elementToPrint = document.getElementsByClassName(className);
        if (elementToPrint && elementToPrint.length > 0) {
            this.print(elementToPrint[0], this.printOpenWindow);
        }
        else {
            console.log('element with id ${className} not found..');
        }
    }
    /**
     * Print Angular TemplateRef or a Component or String
     * @param contentToPrint
     * @example
     * this.printerService.printAngular(this.PrintTemplateTpl);
     */
    printAngular(contentToPrint, context) {
        const nativeEl = this.createComponent(contentToPrint, null, context);
        this.print(nativeEl.nativeElement, this.printOpenWindow);
    }
    /**
     * Print single img
     * @example
     * this.printerService.printImg('assets/bratwurst.jpg');
     */
    printImg(imgSrc) {
        const compRef = this.createComponent(null, imgSrc);
        const openNewWindow = this.printOpenWindow;
        compRef.instance.completed.subscribe(val => {
            compRef.hostView.detectChanges();
            console.log('completed:', val);
            this.print(compRef.location.nativeElement, openNewWindow);
        });
    }
    /**
     * Print an native Element (HTML Element)
     * @param nativeElement
     * @example
     * this.printerService.printHTMLElement(this.PrintComponent.nativeElement);
     */
    printHTMLElement(nativeElement) {
        this.print(nativeElement, this.printOpenWindow);
    }
    /**
     * Create and render component
     * @param contentToRender
     */
    createComponent(contentToRender, imgSrc, context) {
        // this.viewContainerRef.clear();
        const factory = this.resolver.resolveComponentFactory(NgxPrinterComponent);
        let componentRef;
        if (contentToRender) {
            if (context === undefined) {
                context = null;
            }
            const ngContent = this.resolveNgContent(contentToRender, context);
            componentRef = factory.create(this.injector, ngContent); // this.viewContainerRef.createComponent(factory);
        }
        else {
            componentRef = factory.create(this.injector);
        }
        componentRef.instance.renderClass = this.renderClass;
        if (imgSrc) {
            componentRef.instance.imgSrc = imgSrc;
            return componentRef;
        }
        componentRef.hostView.detectChanges();
        return componentRef.location; // location is native element
    }
    /**
     * Main print function
     * @param printContent
     */
    print(printContent, printOpenWindow) {
        if (printOpenWindow === true) {
            const printContentClone = document.importNode(printContent, true); // printContent.cloneNode(true);
            this.printInNewWindow(printContentClone);
        }
        if (printOpenWindow === false) {
            const printContentClone = document.importNode(printContent, true); // printContent.cloneNode(true);
            const nativeEl = this.createComponent(printContentClone).nativeElement;
            this.openNgxPrinter = nativeEl;
            document.body.appendChild(this.openNgxPrinter);
            // window.scrollTo(0, 0);
            this.getAppRoot();
            if (this.appRoot) {
                this.appRoot.style.display = 'none';
            }
            this.printCurrentWindow();
        }
    }
    /**
     * Print using a new window / tab
     * @param divToPrint
     */
    printInNewWindow(divToPrint) {
        const printWindow = window.open('', 'PRINT');
        const title = document.title;
        printWindow.document.write('<HTML><HEAD><TITLE>' + title + '</TITLE></HEAD><BODY></BODY></HTML>');
        // printWindow.document.write(document.documentElement.innerHTML);
        const printWindowDoc = printWindow.document;
        this.copyCss(printWindowDoc);
        printWindowDoc.body.style.margin = '0 0';
        printWindowDoc.body.appendChild(divToPrint);
        printWindow.document.close();
        // printWindowDoc.body.innerHTML = divToPrint.outerHTML;
        setTimeout(() => this.printTabWindow(printWindow, printWindowDoc), this.timeToWaitRender);
    }
    /**
     * Copy Css links to new page
     * @param printWindow
     */
    copyCss(printWindowDoc) {
        const links = document.querySelectorAll('link');
        const styles = document.querySelectorAll('style');
        const targetHead = printWindowDoc.getElementsByTagName('head')[0];
        links.forEach(link => {
            targetHead.appendChild(document.importNode(link, true));
        });
        styles.forEach(style => {
            targetHead.appendChild(document.importNode(style, true));
        });
    }
    /**
     * Print window in new tab
     */
    printTabWindow(printWindow, printWindowDoc) {
        if (this.printPreviewOnly) {
            return;
        }
        this.registerPrintEvent(printWindow, true);
        this.printWindowOpen.next(true);
        printWindow.focus(); // necessary for IE >= 10*/
        if (printWindowDoc.execCommand('print') === false) {
            printWindow.print();
        }
    }
    /**
     * Print the whole current window
     */
    printCurrentWindow() {
        if (this.printPreviewOnly) {
            return;
        }
        this.registerPrintEvent(window, false);
        setTimeout(() => {
            this.printWindowOpen.next(true);
            if (document.execCommand('print') === false) {
                window.print();
            }
        }, this.timeToWaitRender);
    }
    /**
     * Listen to print event of window
     * @param printWindow
     */
    registerPrintEvent(printWindow, printWithOpenInNewWindow) {
        const that = this;
        printWindow.focus(); // necessary for IE >= 10*/
        if (that.eventadded[printWindow.name]) {
            return;
        }
        printWindow.addEventListener('afterprint', () => {
            this.eventadded[printWindow.name] = true;
            // console.log('afterprint');
            if (printWithOpenInNewWindow) {
                that.eventadded[printWindow.name] = false;
            }
            that.cleanUp(printWindow, printWithOpenInNewWindow);
            that.printWindowOpen.next(false);
        });
    }
    /**
     * Close tab or clean up dom
     * @internal
     */
    cleanUp(printWindow, printOpenWindow) {
        if (printOpenWindow === true) {
            console.log('close print window');
            printWindow.close();
            setTimeout(() => {
                printWindow.close();
            }, 20);
        }
        if (printOpenWindow === false) {
            if (!this.openNgxPrinter) {
                return;
            }
            if (document.body.getElementsByTagName('ngx-printer').length === 0) {
                return;
            }
            if (this.appRoot) {
                if (this.appRootDislaySetting !== '') {
                    this.appRoot.style.display = this.appRootDislaySetting;
                }
                else {
                    this.appRoot.style.display = '';
                }
            }
            document.body.removeChild(this.openNgxPrinter);
            this.openNgxPrinter = null;
        }
    }
    /**
     * Create node or angular component
     * @param content
     * @internal
     */
    resolveNgContent(content, context) {
        if (typeof content === 'string') {
            const element = document.createTextNode(content);
            return [[element]];
        }
        if (content instanceof TemplateRef) {
            const viewRef = content.createEmbeddedView(context);
            viewRef.detectChanges();
            return [viewRef.rootNodes];
        }
        if (content instanceof HTMLElement) {
            return [[content]];
        }
        /** Otherwise it's a component */
        const factory = this.resolver.resolveComponentFactory(content);
        const componentRef = factory.create(this.injector);
        componentRef.changeDetectorRef.detectChanges();
        return [[componentRef.location.nativeElement]];
    }
    /**
     * Search for Angular App Root
     * @internal
     */
    getAppRoot() {
        const appRoot = document.body.getElementsByTagName(this.appRootName);
        if (appRoot.length === 0) {
            return null;
        }
        else {
            this.appRoot = appRoot[0];
            this.appRootDislaySetting = this.appRoot.style.display;
        }
    }
    /**
     * Add a new item to print
     * Used by directive
     * @internal
     * @param newPrintItem  HTML id
     */
    addPrintItem(newPrintItem) {
        const tmpItems = this._printItems.getValue();
        tmpItems.push(newPrintItem);
        this._printItems.next(tmpItems);
    }
    /**
     * Delete a print item from service
     * Used by directive
     * @internal
     * @param idOfItemToRemove
     */
    removePrintItem(idOfItemToRemove) {
        const tmpItems = this._printItems.getValue();
        const newIitems = tmpItems.filter(item => item.id !== idOfItemToRemove);
        this._printItems.next(newIitems);
    }
    /**
   * Gets a single print item from service
   * Used by directive
   * @internal
   * @param idOfItemToFind
   */
    getPrintItem(idOfItemToRemove) {
        const tmpItems = this._printItems.getValue();
        const foundItem = tmpItems.find(item => item.id === idOfItemToRemove);
        return foundItem;
    }
    /**
     * Print a print Item
     * @param printItemToPrint
     */
    printPrintItem(printItemToPrint) {
        this.printHTMLElement(printItemToPrint.nativeElement);
    }
    /**
     * Print al list of print Items one after the other
     * @param printItemToPrint
     */
    printPrintItems(printItemsToPrint, className) {
        const newDiv = document.createElement('div');
        if (className) {
            newDiv.classList.add(className);
        }
        else {
            newDiv.style.display = 'flex';
            newDiv.style.flexDirection = 'column';
        }
        printItemsToPrint.forEach(element => {
            newDiv.appendChild(element.nativeElement.cloneNode(true));
        });
        this.printHTMLElement(newDiv);
    }
}
NgxPrinterService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxPrinterService_Factory() { return new NgxPrinterService(i0.ɵɵinject(i1.PrintServiceConfig, 8), i0.ɵɵinject(i0.ComponentFactoryResolver), i0.ɵɵinject(i0.INJECTOR)); }, token: NgxPrinterService, providedIn: "root" });
NgxPrinterService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
NgxPrinterService.ctorParameters = () => [
    { type: PrintServiceConfig, decorators: [{ type: Optional }] },
    { type: ComponentFactoryResolver },
    { type: Injector }
];
