// ngx-printer/esm2015/lib/ngx-printer.module.js
import { NgModule } from '@angular/core';
import { NgxPrinterComponent } from './ngx-printer.component';
import { PrintServiceConfig } from './print-service-config';
import { PrintItemDirective } from './print-item.directive';
import { PrintItemButtonDirective } from './print-item-button.directive';
import { PrintItemMarkerDirective } from './print-item-marker.directive';
export class NgxPrinterModule {
    static forRoot(config) {
        return {
            ngModule: NgxPrinterModule,
            providers: [{ provide: PrintServiceConfig, useValue: config }]
        };
    }
}
NgxPrinterModule.decorators = [
    { type: NgModule, args: [{
                declarations: [NgxPrinterComponent, PrintItemDirective, PrintItemButtonDirective, PrintItemMarkerDirective],
                imports: [],
                exports: [NgxPrinterComponent, PrintItemDirective, PrintItemButtonDirective, PrintItemMarkerDirective],
                entryComponents: [NgxPrinterComponent]
            },] }
];
