// ngx-printer/esm2015/lib/ngx-printer.component.js
import { Component, Renderer2, ElementRef, Output, EventEmitter } from '@angular/core';
/**
 * Component used to render content when printed to current window
 */
export class NgxPrinterComponent {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this._renderClass = 'default';
        /**
         * Display single image
         */
        this._imgSrc = 'default';
        this.completed = new EventEmitter();
    }
    get renderClass() {
        return this._renderClass;
    }
    set renderClass(value) {
        this._renderClass = value;
        this.setCustomClass();
    }
    get imgSrc() {
        return this._imgSrc;
    }
    set imgSrc(value) {
        this._imgSrc = value;
        this.addImage(this._imgSrc);
    }
    ngOnInit() { }
    /**
     * Attach custom class to element
     */
    setCustomClass() {
        const natElement = this.elementRef.nativeElement;
        this.renderer.removeClass(natElement, 'default');
        this.renderer.addClass(natElement, this._renderClass);
    }
    /**
     * Add custom image
     * @param source
     */
    addImage(source) {
        const natElement = this.elementRef.nativeElement;
        const newImgElement = this.renderer.createElement('img');
        this.renderer.setAttribute(newImgElement, 'src', source);
        this.renderer.listen(newImgElement, 'load', (evt) => {
            console.log('loading completed', evt);
            this.completed.emit(true);
        });
        this.renderer.appendChild(natElement, newImgElement);
    }
}
NgxPrinterComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-printer',
                template: `
    <ng-content></ng-content>
  `,
                styles: [":host.default{background-color:#fff;height:100%;left:0;margin:0;position:fixed;top:0;width:100%;z-index:1000000}"]
            },] }
];
NgxPrinterComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
NgxPrinterComponent.propDecorators = {
    completed: [{ type: Output }]
};
