// ngx-printer/esm2015/lib/print-item-button.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import { NgxPrinterService } from './ngx-printer.service';
/**
 * Directly add function to a button to print an item
 */
export class PrintItemButtonDirective {
    constructor(el, printerService) {
        this.el = el;
        this.printerService = printerService;
        /**
         * Id of print item to be printed
         */
        this.printItemId = '';
        /**
         * html id of div to be printed
         */
        this.divID = '';
        /**
         * print item by class name
         */
        this.className = '';
        /**
         * print current window
         */
        this.printWindow = 'false';
    }
    ngOnInit() {
        if (this.el.nativeElement && this.checkInputs()) {
            this.el.nativeElement.addEventListener('click', () => {
                if (this.printItemId !== '') {
                    this.prinPrintItem();
                }
                if (this.divID !== '') {
                    this.printerService.printDiv(this.divID);
                }
                if (this.className !== '') {
                    this.printerService.printByClassName(this.className);
                }
                if (this.printWindow !== 'false') {
                    this.printerService.printCurrentWindow();
                }
            });
        }
    }
    /**
     * Check if at least one property is set
     */
    checkInputs() {
        const check = !(this.printWindow === 'false' && this.printItemId === '' && this.divID === '' && this.className === '');
        return check;
    }
    /**
     * print item from print items
     */
    prinPrintItem() {
        const itemToPrint = this.printerService.getPrintItem(this.printItemId);
        if (itemToPrint) {
            this.printerService.printPrintItem(itemToPrint);
        }
    }
}
PrintItemButtonDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxPrintItemButton]'
            },] }
];
PrintItemButtonDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgxPrinterService }
];
PrintItemButtonDirective.propDecorators = {
    printItemId: [{ type: Input }],
    divID: [{ type: Input }],
    className: [{ type: Input }],
    printWindow: [{ type: Input }]
};
