// ngx-printer/esm2015/lib/print-item.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import { NgxPrinterService } from './ngx-printer.service';
import { PrintItem } from './print-item';
/**
 * A directive to mark and store an HTML-Element as an item which
 * can be printed
 * An id has to be set
 */
export class PrintItemDirective {
    constructor(el, printerService) {
        this.el = el;
        this.printerService = printerService;
    }
    ngOnInit() {
        if (this.el.nativeElement.id) {
            const tmpPrintItem = new PrintItem();
            tmpPrintItem.id = this.el.nativeElement.id;
            tmpPrintItem.nativeElement = this.el.nativeElement;
            this.printerService.addPrintItem(tmpPrintItem);
        }
    }
    ngOnDestroy() {
        if (this.el.nativeElement.id) {
            this.printerService.removePrintItem(this.el.nativeElement.id);
        }
    }
}
PrintItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxPrintItem]'
            },] }
];
PrintItemDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgxPrinterService }
];
PrintItemDirective.propDecorators = {
    printName: [{ type: Input }]
};
